contributor-docs/adrs/adr-003-prop-norms.md

```diff
@@ -0,0 +1,70 @@
+# ADR 003: Prop norms in Primer React Components
+
+## Status
+
+Proposed
+
+## Context
+
+Today we have:
+
+- Implicit conventions not documented anywhere but consistently reflected in our code (e.g., the type of the `sx` prop)
+- Explicit plans to change some of those (e.g., the deprecation of styled-system props)
+- Inconsistencies in our implementation (e.g., when components accept a `ref` prop)
+
+This ADR aims to unify conversations about prop APIs, codify our decisions, and sequence the work to get there.
+
+## Decision
+
+### 🟢 `sx`
+
+All components that ultimately render to the DOM should accept an `sx` prop.
+
+The `sx` prop (of type `SystemStyleObject`) should generally set styles for the root HTML element rendered by the component. An exception would be components like `Dialog`, whose outermost HTML element is a backdrop. In that case, it would be appropriate for `sx` styles to apply to child of the backdrop that is more likely to need styling overrides.
+
+It may make sense to have an additional prop to style a descendant element (e.g., `Dialog` could accept a `backdropSx` prop).
+
+### 🟢 `ref`
+
+All components that ultimately render to the DOM should accept a `ref` prop. That `ref` prop should be passed to the root HTMLElement rendered by the component.
```

Comment:
Our `TextInput` component provides an interesting edge case for this rule because the `<input>` element is nested inside a `<span>`: https://github.com/primer/react/blob/main/src/TextInput.tsx#L156-L171

Nesting the `input` inside a `span` allows us to render an icon inside the input:

![CleanShot 2021-10-01 at 17 09 25](https://user-images.githubusercontent.com/4608155/135697265-b7328c47-3523-4a34-adfc-33d87b60813c.png)


However, this rule would lead me to pass the `ref` to the wrapping `<span>` but as a consumer, I'd expect the `ref` to be passed to the `<input>` element.

I don't necessarily think we need to update this rule though. Maybe the "correct" solution is adopt a more composable API for TextInput that would avoid this problem entirely (similar to Chakra UI):

![CleanShot 2021-10-01 at 17 11 44](https://user-images.githubusercontent.com/4608155/135697343-2d444544-f986-47f6-826d-250d9d467b68.png)

https://chakra-ui.com/docs/form/input#add-elements-inside-input

Reply:
Oooh yeah, I think this is a good argument for @siddharthkp's work on `children`-based APIs.

For this ADR, I might just relax the language to say:

> That `ref` prop should most often be passed to the root HTMLElement rendered by the component, although sometimes it may make more sense to pass it to a descendent node.